staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go

```diff
@@ -152,21 +153,27 @@ func validateStructuralInvariants(s *Structural, lvl level, fldPath *field.Path)
 	}
 	if metadata, found := s.Properties["metadata"]; found && lvl == rootLevel {
 		// metadata is a shallow copy. We can mutate it.
-		_, foundName := metadata.Properties["name"]
-		_, foundGenerateName := metadata.Properties["generateName"]
-		if foundName && foundGenerateName && len(metadata.Properties) == 2 {
-			metadata.Properties = nil
-		} else if (foundName || foundGenerateName) && len(metadata.Properties) == 1 {
-			metadata.Properties = nil
+		invalidProperties := make([]string, 0)
+		for property, _ := range metadata.Properties {
+			if property != "name" && property != "generateName" {
+				invalidProperties = append(invalidProperties, property)
+			}
 		}
+		if len(invalidProperties) > 0 {
+			errMsg := fmt.Sprintf("must not specify any property other than name and generateName: %s", strings.Join(invalidProperties, ","))
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("properties").Key("metadata").Child("properties"), errMsg))
+		}
+	}
+	if metadata, found := s.Properties["metadata"]; found && lvl == rootLevel {
 		metadata.Type = ""
+		metadata.Properties = nil
 		metadata.Default.Object = nil // this is checked in API validation (and also tested)
 		if metadata.ValueValidation == nil {
 			metadata.ValueValidation = &ValueValidation{}
 		}
 		if !reflect.DeepEqual(metadata, Structural{ValueValidation: &ValueValidation{}}) {
 			// TODO: this is actually a field.Invalid error, but we cannot do JSON serialization of metadata here to get a proper message
-			allErrs = append(allErrs, field.Forbidden(fldPath.Child("properties").Key("metadata"), "must not specify anything other than name and generateName, but metadata is implicitly specified"))
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("properties").Key("metadata"), "must not specify anything other than type and properties, but metadata is implicitly specified"))
```

Comment:
this needs tests before this (potentially functionally) change to prove that this PR does not break the existing API.

Reply:
done